### PR TITLE
Interpreter: de-duplicate code

### PR DESCRIPTION
**Pull request by tautschnig**
evaluate_address(byte_extract) and evaluate(byte_extract) had almost the same
implementation.

This commit previously was part of #2165, but I suspect it is somewhat broken as #2165 previously broke test-gen. Awaiting feedback from the automatic reviewer.

### CODE REVIEW

**Review by kroening — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 54ed5888e2d9d54b5463baf4e749842946672cc6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85123929

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 9bae4adfc955f0ce240bfc40e0c5c18e2f4ef7a7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86270759

**Comment by tautschnig** [comment]
We do not have any tests of the interpreter in the code base (which is a shame), and AFAIK this changed caused some trouble. Assigning to @peterschrammel with the hope to receive some guidance, but if anyone else could contribute tests of the interpreter that would be much appreciated.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: f36c08cb6a56c48a6ee20f6fb40d8968c21996da).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102126796
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

File: src/goto-programs/interpreter_evaluate.cpp [hunk 1]
```diff
@@ -885,42 +885,11 @@ void interpretert::evaluate(
     }
     return;
   }
-  else if(expr.id()==ID_byte_extract_little_endian ||
-          expr.id()==ID_byte_extract_big_endian)
-  {
-    if(expr.operands().size()!=2)
```
Comment (thk123) [inline src/goto-programs/interpreter_evaluate.cpp]: Just noting this sanity check on byte_extract is being deleted.

File: src/goto-programs/interpreter_evaluate.cpp [hunk 2]
```diff
@@ -885,42 +885,11 @@ void interpretert::evaluate(
...
-    evaluate(expr.op1(), extract_offset);
-    mp_vectort extract_from;
-    evaluate(expr.op0(), extract_from);
-    if(extract_offset.size()==1 && extract_from.size()!=0)
-    {
-      const typet &target_type=expr.type();
-      mp_integer memory_offset;
-      // If memory offset is found (which should normally be the case)
-      // extract the corresponding data from the appropriate memory location
-      if(!byte_offset_to_memory_offset(
```
Comment (thk123) [inline src/goto-programs/interpreter_evaluate.cpp:876]: You remove this check. No idea what it does, but it should probably be removed in a separate commit to make the move clearer.

File: src/goto-programs/interpreter_evaluate.cpp [hunk 3]
```diff
@@ -959,16 +928,43 @@ void interpretert::evaluate(
...
+          target_type_leaves > 0)
+        {
+          dest.insert(
+            dest.end(),
+            extract_from.begin() + numeric_cast_v<std::size_t>(memory_offset),
+            extract_from.begin() +
+              numeric_cast_v<std::size_t>(memory_offset + target_type_leaves));
+          return;
+        }
+        // we fail
```
Comment (thk123) [inline src/goto-programs/interpreter_evaluate.cpp:931]: This comment doesn't seem quite right - we don't fail - we carry on to the next if - perhaps a throw here, or a return?

File: src/goto-programs/interpreter_evaluate.cpp [hunk 4]
```diff
@@ -959,16 +928,43 @@ void interpretert::evaluate(
...
+      }
       if(!unbounded_size(expr.type()))
       {
         dest.resize(numeric_cast_v<std::size_t>(get_size(expr.type())));
         read(address, dest);
+        return;
```
Comment (thk123) [inline src/goto-programs/interpreter_evaluate.cpp:937]: This return is redundant. Probably an indication that just pulling out this whole branch into a function might make it a little easier to follow.

**Review by thk123 — COMMENTED** [review]
In principle looks fine (And a preliminary run of TG looks OK too) - but in the interest of helping anyone who ends up git blaming an obsecure bug this introduces could the removal of calling `byte_offset_to_memory_offset` be some how factored into a separate commit.

**Review by martin-cs — APPROVED** [review]
Am fine with the changes but ... can we have a test case because the interpreter is dangerously under tested.
